File: leetcode/0072.Edit-Distance/EditDistance.go
package main

import "fmt"

func mins(value int, values ...int) int {
	for _, v := range values {
		if v < value {
			value = v
		}
	}
	return value
}

func levRecur(word1 string, word2 string, memo map[[2]string]int) int {

	if l1, l2 := len(word1), len(word2); l1 == 0 || l2 == 0 {
		if l1 > l2 {
			return l1
		}
		return l2
	}
	if val, ok := memo[[2]string{word1, word2}]; ok {
		return val
	}
	if word1[0] == word2[0] {
		return levRecur(word1[1:], word2[1:], memo)
	}
	dist := 1 + mins(
		levRecur(word1[1:], word2[1:], memo),
		levRecur(word1[1:], word2, memo),
		levRecur(word1, word2[1:], memo),
	)
	memo[[2]string{word1, word2}] = dist
	return dist
}

// Runtime: 40 ms, faster than 10.57% of Go online submissions for Edit Distance.
func minDistance(word1 string, word2 string) int {
	memo := make(map[[2]string]int)
	return levRecur(word1, word2, memo)
}

func popQueue(queue *[][3]int) [3]int {
	tuple := (*queue)[0]
	*queue = (*queue)[1:]
	return tuple
}

func pushQueue(queue *[][3]int, tuples ...[3]int) {
	*queue = append(*queue, tuples...)
}

// Runtime: 8 ms, faster than 43.90% of Go online submissions for Edit Distance.
func minDistanceQueue(word1 string, word2 string) int {
	l1, l2 := len(word1), len(word2)
	queue := [][3]int{{0, 0, 0}}
	isVisited := make(map[[2]int]bool)
	for {
		tuple := popQueue(&queue)
		p1, p2, dist := tuple[0], tuple[1], tuple[2]
		if _, ok := isVisited[[2]int{p1, p2}]; ok {
			continue
		}
		isVisited[[2]int{p1, p2}] = true
		for p1 < l1 && p2 < l2 && word1[p1] == word2[p2] {
			p1++
			p2++
		}
		if p1 >= l1 && p2 >= l2 {
			return dist
		}
		dist++
		pushQueue(
			&queue,
			[3]int{p1 + 1, p2 + 1, dist},
			[3]int{p1, p2 + 1, dist},
			[3]int{p1 + 1, p2, dist},
		)
	}
}

// Runtime: 0 ms, faster than 100.00% of Go online submissions for Edit Distance.
func minDistanceDP(word1 string, word2 string) int {
	l1, l2 := len(word1), len(word2)

	// creat DP Table
	dp := make([][]int, l1+1)
	for i := range dp {
		dp[i] = make([]int, l2+1)
	}
	for i := range dp[0] {
		dp[0][i] = i
	}
	for i := range dp {
		dp[i][0] = i
	}

	for i := 1; i < l1+1; i++ {
		for j := 1; j < l2+1; j++ {
			if word1[i-1] == word2[j-1] {
				dp[i][j] = dp[i-1][j-1]
			} else {
				dp[i][j] = 1 + mins(
					dp[i-1][j-1],
					dp[i-1][j],
					dp[i][j-1],
				)
			}
		}
	}
	return dp[l1][l2]
}

func main() {
	fmt.Println(minDistance("kitten", "sitting"))
	fmt.Println(minDistanceQueue("kitten", "sitting"))
	fmt.Println(minDistanceDP("kitten", "sitting"))
}

File: leetcode/0072.Edit-Distance/EditDistance.py
from functools import lru_cache
from collections import deque

class Solution:

    # Runtime: 240 ms, faster than 19.64% of Python3 online submissions for Edit Distance.
    def minDistance(self, word1: str, word2: str) -> int:
        def lev(a, b, memo={}):
            if not a or not b:
                return max(len(a), len(b))
            if (a, b) in memo:
                return memo[(a, b)]
            if a[0] == b[0]:
                return lev(a[1:], b[1:])
            
            dist = 1 + min(
                lev(a[1:], b[1:]),  # substitution
                lev(a[0:], b[1:]),  # insert or delete
                lev(a[1:], b[0:]),  # insert or delete
            )
            memo[(a, b)] = dist
            return dist
        
        return lev(word1, word2)

    # Runtime: 244 ms, faster than 18.60% of Python3 online submissions for Edit Distance.
    def minDistanceV2(self, word1: str, word2: str) -> int:
        memo = {}
        def lev(a, b, memo):
            if not a or not b:
                return max(len(a), len(b))
            if (a, b) in memo:
                return memo[(a, b)]
            if a[0] == b[0]:
                return lev(a[1:], b[1:], memo)
            
            dist = 1 + min(
                lev(a[1:], b[1:], memo),  # substitution
                lev(a[0:], b[1:], memo),  # insert or delete
                lev(a[1:], b[0:], memo),  # insert or delete
            )
            memo[(a, b)] = dist
            return dist
        return lev(word1, word2, memo)

    # Runtime: 96 ms, faster than 95.57% of Python3 online submissions for Edit Distance.
    def minDistanceV3(self, word1: str, word2: str) -> int:

        @lru_cache(maxsize=None)
        def lev(a, b):
            if not a or not b:
                return max(len(a), len(b))
            if a[0] == b[0]:
                return lev(a[1:], b[1:])
            
            return 1 + min(
                lev(a[1:], b[1:]),  # substitution
                lev(a[0:], b[1:]),  # insert or delete
                lev(a[1:], b[0:]),  # insert or delete
            )
        return lev(word1, word2)

    # Runtime: 96 ms, faster than 95.57% of Python3 online submissions for Edit Distance.
    def minDistanceV3_1(self, word1: str, word2: str) -> int:

        @lru_cache(maxsize=None)
        def lev(a, b):
            if not a or not b:
                return max(len(a), len(b))
            if a[0] == b[0]:
                return lev(a[1:], b[1:])
            l1, l2 = len(a), len(b)
            p1, p2 = 0, 0
            while p1 < l1 and p2 < l2 and a[p1] == b[p2]:
                p1 += 1
                p2 += 1
            
            return 1 + min(
                lev(a[p1+1:], b[p2+1:]),  # substitution
                lev(a[p1:], b[p2+1:]),  # insert or delete
                lev(a[p1 + 1:], b[p2:]),  # insert or delete
            )
        return lev(word1, word2)


    # Runtime: 52 ms, faster than 99.78% of Python3 online submissions for Edit Distance.
    def minDistanceV4(self, word1: str, word2: str) -> int:
        l1, l2 = len(word1), len(word2)
        visited = set()
        queue = [(0, 0, 0)]
        while queue:
            p1, p2, dist = queue.pop(0)
            if (p1, p2) in visited or p1 > l1 or p2 > l2:
                continue
            if word1[p1:] == word2[p2:]:
                return dist

            visited.add((p1, p2))
            while p1 < l1 and p2 < l2 and word1[p1] == word2[p2]:
                p1 += 1
                p2 += 1
            
            queue.extend(
                [
                    (p1 +1, p2 +1, dist +1),
                    (p1, p2 +1, dist +1),
                    (p1 +1, p2, dist +1),
                ]
            )

    # Runtime: 40 ms, faster than 100.00% of Python3 online submissions for Edit Distance.
    def minDistanceV5(self, a: str, word2: b) -> int:
        la, lb = len(a), len(b)
        isVisited = set()
        queue = [(0, 0, 0)] # (p1, p2, dist)
        while queue:
            p1, p2, dist = queue.pop(0)
            if (p1, p2) in isVisited:
                continue
            isVisited.add((p1, p2))
            while p1 < la and p2 < lb and a[p1] == b[p2]:
                    p1 += 1
                    p2 += 1
            if p1 >= la and p2 >= lb:
                return dist

            dist += 1
            queue.extend([
                (p1 + 1, p2 + 1, dist),
                (p1, p2 + 1, dist),
                (p1 + 1, p2, dist),
            ])